src/Compilers/CSharp/Portable/Symbols/SubstitutedMethodSymbol.cs

```diff
@@ -38,9 +38,10 @@ internal class SubstitutedMethodSymbol : WrappedMethodSymbol
 
         private int _hashCode; // computed on demand
 
-        internal SubstitutedMethodSymbol(SubstitutedNamedTypeSymbol containingSymbol, MethodSymbol originalDefinition)
+        internal SubstitutedMethodSymbol(NamedTypeSymbol containingSymbol, MethodSymbol originalDefinition)
             : this(containingSymbol, containingSymbol.TypeSubstitution, originalDefinition, constructedFrom: null)
         {
+            Debug.Assert(containingSymbol is SubstitutedNamedTypeSymbol || containingSymbol is SubstitutedErrorTypeSymbol);
```

Comment:
The only reason it is "Unsafe" for this kind of use is because of the inappropriate tight coupling fixed by this PR. It should always be safe to call `method.AsMember` with a type that is a construction of the method's containing type. It wasn't before this PR. It is after this PR. That was clearly a bug. With that bug fixed, the crash is fixed.